lilo/bootheader/man/mkzimage.8
.\" $Id$
.TH mkzimage 8
.SH NAME
\fBmkzimage\fR - create bootimage
.SH SYNOPSIS
\fBmkzimage --vmlinux <ELF binary> --output <zImage>\fR [options]
.SH DESCRIPTION
\fBmkzimage\fR creates a bootimage which can be booted directly from firmware.
It can create images for PowerMac, IBM pSeries, IBM iSeries and PReP systems.
.SH OPTIONS
.TP
.B \-\-vmlinux <ELF-binary>
filename of the vmlinux file to boot. This option is mandatory.
.TP
.B \-\-output <zImage>
filename of the output bootfile. This option is mandatory.
.TP
.B \-\-initrd <ramdisk>
filename of the initrd to use.
.TP
.B \-\-objdir <directory>
directory with the helper scripts and objects files to link. This is useful when building from source. Use \fBmkzimage --objdir `pwd`\fR to use the current directory as base.
.TP
.B \-\-cmdline 'kernel cmdline'
force "kernel cmdline" to be passed to the kernel. Options specified in the OpenFirmware \fBboot-file\fR property are ignored. It is possible to edit the cmdline during boot.
.TP
.B \-\-board <subarch>
build a bootimage suiteable for another PPC subarch.
Recognized options are:
.SP 1
.PD 0
.RS
.TP
.B chrp|rs6k
build an ELF image for CHRP, and add a NOTE section required for IBM RS/6000 and pSeries firmware. This NOTE section will break PowerMacs, do never boot such an image on a Mac!
.TP
.B iseries
append an initrd for IBM iSeries partitions. The normal vmlinux ELF binary will work out of the box. A kernel cmdline can not be appended, the --cmdline option is ignored.
.TP
.B pmac
build an ELF image for PowerMac. The size of the final image must be smaller than 4MB because the firmware can not load files larger than 4MB (4MB - 15kb actually).
.TP
.B pmaccoff
build an XCOFF image for old PowerMacs. They must be booted via network, loading them from disk will likely not work.
.TP
.B prep
build a boot image for PReP systems.
.RE
.PD
.TP
.B \-\-tmp <tempdir>
use \fBtempdir\fR as directory for temporary files during linking.
.SH AUTHOR
Olaf Hering <olh@suse.de>
.SH SEE ALSO
.BR mkzimage_cmdline (8)


yaboot/CHANGES
August 6. 2001 - 1.2.3

 - Fix again system.map loading. That also fix an old pending
   bug we had where yaboot could pass random values for system.map,
   causing the kernel to mark random pages reserved.
 - MD5 passwords (Ethan Benson)
 - Fix F80 & remove obsolete ppc64 stuff (Peter Bergner)
 - Add reiserfs support (Jeff Mahoney)

July 28. 2001 - 1.2.2

 - Fix to system.map loading

May 05. 2001 - 1.2

 - Add support for large ramdisks (Edward Swarthout)
 - Added password protection, ported from silo. (Ethan Benson)
 - Added silo's single-key feature. (Ethan Benson)

Feb 18. 2001 - 1.1.1

 - Added a few Makefile cleanup & fixes (Ethan Benson, Peter Bergner)
 - Some image path fixes (Peter Bergner)
 - Don't call setup_display on CHRP
 - Fix color stuff



lilo/bootheader/scripts/mkzimage.sh
#!/bin/bash
# $Id$
set -e
# set -x

export LANG=C
export LC_ALL=C
obj_dir=/lib/lilo

vmlinux=
initrd=
tmp=
outputfile=
cmdline=
#
options=
#
board_type=guess
kernel_type=guess
zimage_sh=guess

until [ "$#" = "0" ] ; do
	case "$1" in
		--help|-h|--version)
		echo "create a 'zImage' for ppc/ppc64, bootable from Openfirmware prompt"
		echo "Usage: ${0##*/} --vmlinux <ELF binary> --output <zImage>"
		echo "additional options:"
		echo "                    [--objdir <dir>]"
		echo "                    [--initrd <ramdisk.image.gz>]"
		echo "                    [--cmdline <kernelcmdline>]"
		echo "                    [--tmp <tempdir>]"
		echo "                    [--board <subarch>]"
		echo "option --board requires a ppc/ppc64 subarch. It can be one of:"
		echo "chrp|rs6k iseries pmaccoff pmac prep"
		exit 1
		;;
		--board)
		shift
		if [ "$#" = "0" -o "$1" = ""  ] ; then
			exec $0 --help
			exit 1
		fi
		board_type=$1
		shift
		;;
		--cmdline)
		shift
		if [ "$#" = "0" -o "$1" = ""  ] ; then
			exec $0 --help
			exit 1
		fi
		cmdline=$1
		shift
		;;
		--vmlinux)
		shift
		if [ "$#" = "0" -o "$1" = ""  ] ; then
			echo "option --vmlinux requires a filename"
			exit 1
		fi
		vmlinux=$1
		shift
		;;
		--initrd)
		shift
		if [ "$#" = "0"  -o "$1" = "" ] ; then
			echo "option --initrd requires a filename"
			exit 1
		fi
		initrd=$1
		shift
		;;
		--output)
		shift
		if [ "$#" = "0" -o "$1" = "" ] ; then
			echo "option --output requires a filename"
			exit 1
		fi
		output=$1
		shift
		;;
		--objdir)
		shift
		if [ "$#" = "0" -o "$1" = ""  ] ; then
			echo "option --objdir requires a diretory"
			exit 1
		fi
		obj_dir=$1
		shift
		;;
		--tmp)
		shift
		if [ "$#" = "0" -o "$1" = ""  ] ; then
			echo "option --tmp requires a diretory"
			exit 1
		fi
		tmp=$1
		shift
		;;
		*)
		echo "ERROR: unknown option $1"
		exec $0 --help
		;;
	esac
done
if [ -z "$vmlinux" ] ; then
	echo "ERROR: no input file"
	exec $0 --help
	exit 1
fi
if [ -z "$output" ] ; then
	echo "ERROR: no output file"
	exec $0 --help
	exit 1
fi
#
case "$(file -Lb $vmlinux)" in
	ELF\ 64-bit*)
		kernel_type=64bit
		;;
	ELF\ 32-bit*)
		kernel_type=32bit
		;;
	*)
		file -b $vmlinux
		echo wrong filetype
		exit 1 
		;;
esac
if [ "$board_type" = "guess" ] ; then
	test -f /proc/cpuinfo || mount -v -n -t proc proc /proc
	while read line; do
		case "$line" in
		  *MacRISC*)	board_type="pmac" ;;
		  *CHRP*)  	board_type="chrp";;
		  *PReP*)  	board_type="prep";;
		  *iSeries*)    board_type="iseries";;
		  pmac-generation*)
			set -- $line
			board_type=$3
			;;
		esac
	done < /proc/cpuinfo
fi
if [ "$board_type" = "guess" ] ; then
	test -f /proc/cpuinfo || mount -v -n -t proc proc /proc
	line=`cat < /proc/device-tree/model`
	case "$line" in
		Momentum,Maple-D)	board_type="chrp" ;;
		Momentum,Maple-L)	board_type="chrp" ;;
		*) echo "board type '$line' not recognized, specify one with --board" ;;
	esac
fi
case "$kernel_type" in
	64bit)
	case "$board_type" in
			chrp|pseries|rs6k)
			zimage_sh=make_zimage_chrp.sh
			;;
			iseries)
			zimage_sh=make_zimage_iseries.sh
			;;
			pmac|NewWorld)
			zimage_sh=make_zimage_pmac_newworld.sh
			;;
			*)
			echo "ERROR: boardtype \"$board_type\" not supported as 64bit"
			exec $0 --help
			;;
	esac
	;;
	32bit)
	case "$board_type" in
			chrp|rs6k)
			zimage_sh=make_zimage_chrp.sh
			;;
			pmaccoff|OldWorld)
			zimage_sh=make_zimage_pmac_oldworld_coff.sh
			;;
			prep)
			zimage_sh=make_zimage_prep.sh
			;;
			pmac|NewWorld)
			zimage_sh=make_zimage_pmac_newworld.sh
			;;
			*)
			echo "ERROR: boardtype \"$board_type\" not supported as 32bit"
			exec $0 --help
			;;
	esac
	;;
esac
#
echo board_type $board_type kernel_type $kernel_type zimage_sh $zimage_sh

#
#
#
if [ ! -z "$initrd" ] ; then
options="--initrd $initrd"
fi
if [ ! -z "$tmp" ] ; then
options="$options --tmp $tmp"
fi
bash $obj_dir/scripts/$zimage_sh $options --vmlinux "$vmlinux" --output "$output" --objdir "$obj_dir"
if [ ! -z "$cmdline" ] ; then
$obj_dir/utils/mkzimage_cmdline -a 1 -c -s "$cmdline" "$output"
fi
